src/test/kotlin/edu/kit/compiler/parser/SourceRangeTest.kt

```diff
@@ -0,0 +1,214 @@
+     */
+    fun append(newRange: SourceRange) {
+        if (ranges.isEmpty()) {
+            ranges[1] = mutableListOf(newRange)
+        } else {
+            val firstLineWithEnoughSpace = ranges.entries.firstOrNull {
+                it.value.last().last.offset < newRange.start.offset
+            }?.key
+            if (firstLineWithEnoughSpace == null) {
+                ranges[ranges.keys.maxOf { it } + 1] = mutableListOf(newRange)
+            } else {
+                ranges[firstLineWithEnoughSpace]!!.add(newRange)
+            }
+        }
+    }
+
+
+    fun drawRange(offset: Int, start: Int, length: Int): String {
+        return " ".repeat(start - offset - 1) + when (length) {
+            0 -> "-"
+            1 -> "|"
+            else -> "|" + "-".repeat(length - 2) + "|"
+        }
+    }
+
+    fun format(): String =
+        ranges.map {
+            var lastExtent = 0
+            it.value.joinToString("") { range ->
+                val res = if (range.start.line < currentLine && range.last.line == currentLine) {
+                    "<-|"
+                } else if (range.start.line == currentLine && range.last.line == currentLine) {
+                    drawRange(lastExtent, range.start.column, range.length)
+                } else if (range.start.line == currentLine) {
+                    "|->"
+                } else {
+                    "????"
+                }
+                lastExtent = range.last.column
+                res
+            }
+        }.joinToString("\n") { "#${it.substring(1)}" }
+}
+
+class SourceRangeTest {
+    class SourceRangeCollector(val ranges: MutableList<SourceRange> = mutableListOf()) : AbstractAstVisitor() {
+        override fun visitSourceRange(sourceRange: SourceRange): SourceRange {
+            ranges.add(sourceRange)
+            return super.visitSourceRange(sourceRange)
+        }
+    }
+
+    fun <T> expectRanges(code: String, runParser: Parser.() -> T, runVisitor: SourceRangeCollector.(T) -> Unit) {
+
+        val (expectedRanges, source) = code.lineSequence().partition {
+            it.startsWith("#")
+        }
+        println("source: $source")
+        val parsed = withParser(source.joinToString("\n")) { runParser() }
+        val sourceRangeCollector = SourceRangeCollector()
+        println("parsed:    $parsed")
+        sourceRangeCollector.runVisitor(parsed)
+        val actualRanges = sourceRangeCollector.ranges
+
+        val actualSortedRanges = mutableListOf<List<SourceRange>>()
+
+        val lastLine = actualRanges.map { it.last.line }.maxOrNull()
+        val rangesByLines = (0..(lastLine ?: 0)).map { currentLine ->
+            actualRanges.filter { it.start.line <= currentLine && it.last.line >= currentLine }
+                .sortedBy { it.length }
+                .sortedBy { it.start.column }
+        }.withIndex()
+            .associate { Pair(it.index, it.value) }
```

Comment:
Nicht nötig, aber bei sowas find ich die Syntax `.associate { it.index to it.value }` ganz schön^^

Reply:
ups :) Stimmt `to` gibt's ja auch :D

Und stimmt `map, withIndex, associate` geht einfacher und schöner mit nur `associate`